tiledb/sm/array/test/unit_consistency.h

```diff
@@ -0,0 +1,139 @@
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines common test functions for class Array and its supporting
+ * classes.
+ */
+
+#ifndef TILEDB_UNIT_CONSISTENCY_H
+#define TILEDB_UNIT_CONSISTENCY_H
+
+#include <catch.hpp>
+#include <iostream>
+
+#include "../array.h"
+#include "../consistency.h"
+#include "tiledb/sm/array_schema/dimension.h"
+#include "tiledb/sm/enums/array_type.h"
+#include "tiledb/sm/enums/encryption_type.h"
+#include "tiledb/sm/enums/layout.h"
+
+using namespace tiledb;
+using namespace tiledb::common;
+using namespace tiledb::sm;
+
+class ConsistencySentry;
+
+/* ************************************* */
+/*  class WhiteboxConsistencyController  */
+/* ************************************* */
+
+namespace tiledb::sm {
+
+using entry_type = std::multimap<const URI, Array&>::const_iterator;
+
+class WhiteboxConsistencyController : public ConsistencyController {
+ public:
+  WhiteboxConsistencyController() = default;
+  ~WhiteboxConsistencyController() = default;
+
+  entry_type register_array(const URI uri, Array& array) {
+    return ConsistencyController::register_array(uri, array);
+  }
+
+  void deregister_array(entry_type entry) {
+    ConsistencyController::deregister_array(entry);
+  }
+
+  ConsistencySentry make_sentry(const URI uri, Array& array) {
+    return ConsistencyController::make_sentry(uri, array);
+  }
+
+  bool is_open(const URI uri, Array& array) {
+    return ConsistencyController::is_open(uri, array);
+  }
+
+  bool is_open(const URI uri) {
+    return ConsistencyController::is_open(uri);
+  }
+
+  bool is_element_of(URI uri1, URI uri2) {
+    return ConsistencyController::is_element_of(uri1, uri2);
+  }
+
+  size_t registry_size() {
+    return ConsistencyController::registry_size();
+  }
+
+  tdb_unique_ptr<Array> open_array(const URI uri, StorageManager* sm) {
```

Comment:
I'm using a unique_ptr to allow the function to return _some reference to_ an Array without it going out of scope. That way I can check the registration in the test and close the array before checking the (now empty) registration again. I can move the contents of this function _into_ the test code so I'm not returning a unique_ptr and risking the array going out of scope; I did it this way because it was cleaner.